proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,747 @@
+``case``. Specifically in ``case_p u of {…; C x1 … xn -> …; …}`` Where ``C :: a1:q1 -> … an:qn -> A``,
+Then ``xi`` has multiplicity annotation ``p*qi``. For instance
+
+::
+
+  bar :: (a,b):p -> c
+  bar (x,y) = … -- Since (,) :: a ->. b ->. (a,b), x and y have
+                -- multiplicity p
+
+There are unresolved issue regarding inference (see Unresolved
+questions below for more):
+
+- There is no account of multiplicity inference. A better
+  understanding would make inference more predictable.
+- For ``let`` bindings and ``case`` expressions which are not part of
+  an equation, we want to infer the multiplicity annotation. The
+  process for this is not yet defined.
+
+TODO: subtyping-like situations (first-order)
+TODO: forbid ``case_0`` (and variable substitution, and type application)
+
+Effect and Interactions
+-----------------------
+
+A staple of this proposal is that it does not modify Haskell for those
+who don't want to use it, or don't know of linear types. Even if an
+API exports linear types, they are easy to ignore: just imagine that
+the arrows are regular arrows, it will work as expected.
+
+Linear data types are just regular Haskell type, which means its cheap
+to get interact with existing libraries.
+
+There is one known unpleasant interaction: with
+``-XRebindableSyntax``, ``if u then t else e`` is interpreted as
+``ifThenElse u t e``. Unfortunately, these two construct have
+differrent typing rules when ``t`` and ``e`` have free linear
+variables. Therefore well-typed linearly typed programs can stop
+typing when ``-XRebindableSyntax`` is added.
+
+TODO: view patterns
+Unresolved: view patterns
+Unresolved: ``@`` patterns
+Remark: lazy pattern-matching is only allowed for patterns as
+multiplicity ``ω``.
+TODO: when using a library with linear types and you don't: you won't
+be able to use functions which take linear functions as arguments, but
+first-order functions with linear types will work silently. When using
+a function which require a linear type, you will get an error from ``-XLinearTypes``.
+
+TODO (better phrasing)
+There is a an issue that defining newtypes such as
+
+::
+
+  newtype Unrestricted' a where
+    Unrestricted' :: a -> Unrestricted' a
+
+Because of laziness which could cause linear values not to be
+consumed.
+
+Costs and Drawbacks
+-------------------
+
+TODO: clean header
+Give an estimate on development and maintenance costs. List how this
+effects learnability of the language for novice users. Define and list
+any remaining drawbacks that cannot be resolved.
+
+This proposal tries hard to make the changes invisible to newcomers,
+however, if many libraries start adopting it, the new function types
+will appear in APIs. They can be safely ignore, but they can still be
+considered distracting.
+
+TODO: the arrow type constructor is used very often in GHC's internal,
+so there are many place which need to be handle linearity. It's often
```

Comment:
GHC specific. Is it relevant?

Reply:
I think that for a costs and drawbacks section, GHC specific is legit.